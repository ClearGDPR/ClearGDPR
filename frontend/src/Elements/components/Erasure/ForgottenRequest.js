import React from 'react';
import PropTypes from 'prop-types';

import ModalView from '../Common/Views/Modal';
import styles from '../../theme/ForgottenRequest.scss';

import { inject } from '../../../containers/SubjectContext';

const about = `The data subject shall have the right to request from the controller the erasure of
personal data concerning him or her without undue delay and the controller shall have
the obligation to erase personal data without undue delay where one of the following
grounds applies:`;

class ForgottenRequest extends React.PureComponent {
  state = {
    openModal: false,
    processing: false,
    success: false,
    error: null
  };

  eraseData() {
    this.setState({ processing: true });
    this.props.cg.Subject.eraseData()
      .then(() => {
        this.setState({
          success: true,
          processing: false,
          err: null
        });
        this.props.subject.initNewSession();
      })
      .catch(err => {
        this.setState({
          success: false,
          processing: false,
          error: err
        });
      });
  }

  handleClick(e) {
    e.preventDefault();
    this.setState({
      success: true
    });
  }

  onOpenModal = () => {
    this.setState({
      success: false,
      openModal: true
    });
  };

  onCloseModal = () => {
    this.setState({
      openModal: false
    });
  };

  render() {
    const { processing, success, error } = this.state;

    const Success = () => (
      <div className={styles.success}>
        <h2 className={styles.title}>Your data was erased!</h2>
        <p className={styles.about}>You can still access the log of your information.</p>
      </div>
    );

    const Errors = ({ errors }) => {
      return errors ? <div className={styles.error}>{errors.error}</div> : null;
    };

    return (
      <React.Fragment>
        <button className="button is-primary" onClick={this.onOpenModal}>
          {this.props.options.label}
        </button>
        <ModalView open={this.state.openModal} onClose={this.onCloseModal}>
          {!success ? (
            <div className={styles.container}>
              <div className={styles.full}>
                <h2 className={styles.heading}>Erase Data</h2>
                <h3 className={styles.subheading}>You're about to erase your data</h3>
                <p className={styles.text}>{about}</p>
                <button
                  className={`button is-primary ${processing ? 'is-loading' : ''}`}
                  onClick={this.eraseData.bind(this)}
                >
                  Send Request
                </button>
                <Errors errors={error} />
              </div>
            </div>
          ) : (
            <Success />
          )}
        </ModalView>
      </React.Fragment>
    );
  }
}

ForgottenRequest.propTypes = {
  options: PropTypes.object,
  subject: PropTypes.object,
  cg: PropTypes.object
};

export default inject(ForgottenRequest);
