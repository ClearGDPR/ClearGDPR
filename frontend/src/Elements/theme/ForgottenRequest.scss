@import 'variables';

.container {
  display: flex;
  margin-bottom: 30px;
}
.full {
  max-width: 100%;
  padding-right: 2rem;

  button {
    margin: 10px auto 0;
    background: $water;
  }
}
.left {
  max-width: 100%;
  padding-right: 2rem;
}
.right {
  max-width: 50%;
  padding-top: 60px;
  text-align: center;

  label {
    margin-bottom: 20px;
    font-size: 14px;
    text-align: left;
  }

  button {
    margin: 10px auto 0;
    background: $water;
  }
}
.heading {
  color: #697689;
  font-size: 36px;
  font-weight: 600;
  letter-spacing: 0.6px;
}
.subheading {
  color: #aeb9c9;
  font-size: 24px;
  letter-spacing: 0.43px;
  margin-bottom: 10px;
}
.text {
  color: #2d323a;
  font-size: 16px;
  letter-spacing: 0.4px;
  line-height: 24px;
}
.success {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  min-width: 680px;
  min-height: 350px;
  text-align: center;
}

.title {
  color: #9cf0dc;
  font-size: 36px;
  font-weight: bold;
}

.about {
  margin: 40px 0;
  width: 75%;
}
.error {
  color: red;
  margin-top: 8px;
  font-size: 12px;
}
