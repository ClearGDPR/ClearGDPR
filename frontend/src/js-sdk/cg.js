import 'es6-promises';
import Config from './config';
import Modules from './modules';
import { SDKError } from './common/exceptions';
import { mergeOptions } from './utils/helpers';

/**
 * Creates a new CG ClearGDPR API client
 * @constructor
 * @classdesc The CG SDK class provides methods to manage user's data
 * @param {Object}  options
 * @param {String}  options.apiKey Key used to sign up to the API
 * @param {String}  options.apiUrl URL to the API server without trailing slash
 * @param {String}  options.apiVersion Version of the API, like `v1`
 * @param {String}  options.apiTimeout Timeout used when a Request is made to the API
 * @param {Boolean} options.debug Flag to enable or disable complete error stack trace
 * @see {@link https://docs.cleargdpr.com/api}
 */
export class CG {
  constructor({ apiKey, apiUrl }) {
    // Enforce accessToken option to be sent.
    if (!apiKey) {
      throw new SDKError('API Key is required.');
    }

    if (!apiUrl) {
      throw new SDKError('API URL is required.');
    }

    this._config = mergeOptions(Config, { apiUrl });
    this.setAPIKey(apiKey);
    this.registerModules();
  }

  registerModules() {
    Object.keys(Modules).map(key => {
      const endpoint = new Modules[key](this);
      Object.assign(this, { [key]: endpoint });
    });
  }

  getAccessToken() {
    return this._config.auth;
  }

  setAccessToken(accessToken) {
    if (accessToken) {
      // TODO: Should validate quicker if its a valid JWT
      // TODO: How to verify JWT with our Server + Consumer
      this._setConfigField('auth', `Bearer ${accessToken}`);
    }
  }

  getAPIKey() {
    return this._config.auth;
  }

  setAPIKey(apiKey) {
    if (apiKey) {
      this._setConfigField('apiKey', apiKey);
    }
  }

  getBaseURL() {
    return this._config.apiVersion
      ? `${this._config.apiUrl}/${this._config.apiVersion}`
      : this._config.apiUrl;
  }

  _setConfigField(key, value) {
    this._config[key] = value;
  }
}
