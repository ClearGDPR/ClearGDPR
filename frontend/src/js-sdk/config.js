const Config = {
  auth: null,
  apiKey: process.env.API_KEY,
  apiUrl: process.env.API_URL,
  apiVersion: process.env.API_VERSION,
  apiTimeout: 30,
  debug: false
};

export default Config;
