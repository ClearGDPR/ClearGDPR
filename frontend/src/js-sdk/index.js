/*
 * cg-js-sdk
 * https://www.cleargdpr.com/
 * 
 */
export { CG } from './cg';
