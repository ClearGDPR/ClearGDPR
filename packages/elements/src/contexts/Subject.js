import _ from 'lodash';
import { DATA_ERASED, DATA_STATUS } from '../constants';

export default class Subject {
  constructor(
    cg,
    {
      isFetched = false,
      isLoading = false,
      isErased = false,
      data = null,
      status = null,
      processors = null,
      objection = null,
      shares = null,
      restrictions = null,
      isGuest = true,
      propagateMutation
    }
  ) {
    this.cg = cg;

    this.isFetched = isFetched;
    this.isLoading = isLoading;
    this.isErased = isErased;
    this.data = data;
    this.status = status;
    this.processors = processors;
    this.objection = objection;
    this.shares = shares;
    this.restrictions = restrictions;
    this.isGuest = isGuest;

    this._propagateMutation = () => {
      propagateMutation(new Subject(cg, { ...this, propagateMutation }));
    };
  }

  async fetchData() {
    const { isLoading, isFetched, isGuest } = this;

    if (isLoading || isFetched || isGuest) {
      return;
    }

    this.isLoading = true;

    let data;
    let status;

    try {
      status = await this.cg.Subject.getDataStatus();
    } catch (e) {
      status = null;
    }

    const isErased = !status || status.controller === DATA_ERASED;

    if (!isErased) {
      try {
        data = await this.cg.Subject.accessData();
      } catch (e) {
        data = null;
      }
    }

    this.data = data;
    this.isLoading = false;
    this.isFetched = true;
    this.isErased = isErased;
    this.status = status;

    this._propagateMutation();
  }

  async fetchProcessors() {
    if (this.processors) {
      return;
    }

    let processors;
    try {
      processors = await this.cg.Subject.getProcessors();
    } catch (e) {
      console.log('failure', e);
      return;
    }

    this.processors = processors.map(p => {
      p.enabled = true;

      if (this.status) {
        const s = _.find(this.status.processors, { id: p.id }) || {};
        p.status = s.status ? DATA_STATUS[s.status] : DATA_STATUS.UNCONSENTED;
      }

      return p;
    });

    this._propagateMutation();
  }

  async initiateRectification(requestReason, rectificationPayload) {
    return await this.cg.Subject.initiateRectification(requestReason, rectificationPayload);
  }

  async eraseData() {
    await this.cg.Subject.eraseData();
    this.status.controller = DATA_ERASED;
    this.data = null;
    this.isErased = true;
    this._propagateMutation();
  }

  async fetchObjectionStatus() {
    if (this.objection !== null) {
      return;
    }
    const { objection } = await this.cg.Subject.getObjectionStatus();
    this.objection = objection;
    this._propagateMutation();
  }

  async updateObjection(value) {
    await this.cg.Subject.updateObjection(value);
    this.objection = value;
    this._propagateMutation();
  }

  async fetchDataShares() {
    if (this.shares !== null) {
      return;
    }
    try {
      this.shares = await this.cg.Subject.getDataShares();
    } catch (e) {
      console.log('failure', e);
      return;
    }

    this._propagateMutation();
  }

  async addDataShare(name) {
    try {
      await this.cg.Subject.addDataShare({
        name
      });
    } catch (e) {
      console.log(e);
      return;
    }
    this.shares = null;
    this.fetchDataShares();
  }

  async removeDataShare(id) {
    try {
      await this.cg.Subject.removeDataShare(id);
    } catch (e) {
      console.log(e);
      return;
    }
    this.shares = null;
    this.fetchDataShares();
  }

  async fetchRestrictions() {
    if (this.restrictions !== null) {
      return;
    }
    this.restrictions = await this.cg.Subject.getRestrictions();
    this._propagateMutation();
  }

  async updateRestrictions(restrictions) {
    this.restrictions = restrictions;
    this._propagateMutation();

    try {
      await this.cg.Subject.updateRestrictions(restrictions);
    } catch (e) {
      console.log(e);
    }
  }

  async giveConsent(payload) {
    try {
      await this.cg.Subject.giveConsent(payload);
    } catch (e) {
      console.log('failure', e);
    }
  }

  onEntrance(callback) {
    this.cg.Events.subscribe('auth.setAccessToken', callback);
  }

  clearEntranceListeners() {
    this.cg.Events.clear('auth.setAccessToken');
  }
}
