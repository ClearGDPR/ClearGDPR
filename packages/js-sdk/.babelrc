{
  "presets": [
    [
      "env",
      {
        "targets": {
          "browsers": ["> 1%", "last 2 versions"]
        }
      }
    ]
  ],
  "plugins": [
    [
      "module-resolver",
      {
        "extensions": [".js", ".jsx", ".es", ".es6", ".mjs"],
        "root": ["./src"],
        "alias": {}
      }
    ]
  ]
}
