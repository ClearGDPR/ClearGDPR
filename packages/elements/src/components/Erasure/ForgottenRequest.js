import React from 'react';
import PropTypes from 'prop-types';

import ModalView from '../Common/Views/Modal';
import styles from '../../theme/ForgottenRequest.scss';
import Subject from '../../contexts/Subject';

const about = `The data subject shall have the right to request from the controller the erasure of
personal data concerning him or her without undue delay and the controller shall have
the obligation to erase personal data without undue delay where one of the following
grounds applies:`;

class ForgottenRequest extends React.PureComponent {
  state = {
    openModal: false,
    processing: false,
    error: null
  };

  async eraseData() {
    this.setState({ processing: true });
    try {
      await this.props.subject.eraseData();
    } catch (e) {
      this.setState({
        processing: false,
        error: e
      });
    }
  }

  onOpenModal = () => {
    this.setState({
      openModal: true
    });
  };

  onCloseModal = () => {
    this.setState({
      openModal: false
    });
  };

  render() {
    const { processing, error } = this.state;

    const Errors = ({ errors }) => {
      return errors ? <div className={styles.error}>{errors.error}</div> : null;
    };

    return (
      <React.Fragment>
        <button className="button is-primary" onClick={this.onOpenModal}>
          {this.props.options.label}
        </button>
        <ModalView open={this.state.openModal} onClose={this.onCloseModal}>
          <div className={styles.container}>
            <div className={styles.full}>
              <h2 className={styles.heading}>Erase Data</h2>
              <h3 className={styles.subheading}>You're about to erase your data</h3>
              <p className={styles.text}>{about}</p>
              <button
                className={`button is-primary ${processing ? 'is-loading' : ''}`}
                onClick={this.eraseData.bind(this)}
              >
                Send Request
              </button>
              <Errors errors={error} />
            </div>
          </div>
        </ModalView>
      </React.Fragment>
    );
  }
}

ForgottenRequest.propTypes = {
  options: PropTypes.object,
  subject: PropTypes.instanceOf(Subject)
};

export default ForgottenRequest;
