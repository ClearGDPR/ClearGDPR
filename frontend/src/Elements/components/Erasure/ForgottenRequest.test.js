import React from 'react';
import { CG } from '@cleargdpr/js-sdk';
import { mount } from 'enzyme';
import ForgottenRequest from './ForgottenRequest';
import ModalView from '../Common/Views/Modal';
import Subject from '../../contexts/Subject';

const cg = new CG({
  apiKey: 'test',
  apiUrl: 'test'
});
cg.Subject.eraseData = async () => {};

const subject = new Subject(cg, {
  status: 1,
  data: {},
  initNewSession: () => {}
});

const setup = () => mount(<ForgottenRequest {...{ options: { label: 'Erase' }, subject }} />);

let spy;

describe('ForgottenRequest', () => {
  it('should render correctly', () => {
    const component = setup();
    expect(component).toMatchSnapshot();
  });

  it('should open confirmation modal before erase', async () => {
    const component = setup();
    expect(component.find(ModalView).props().open).toBe(false);
    component.find('button').simulate('click');
    expect(component.find(ModalView).props().open).toBe(true);
  });

  it('should execute `eraseData` after confirmation and sets `processing` state attribute to true', async () => {
    const component = setup();
    component.find('button').simulate('click');
    spy = jest.spyOn(subject, 'eraseData');

    component
      .find(ModalView)
      .find('button.is-primary')
      .simulate('click');

    expect(spy).toHaveBeenCalled();
  });

  afterEach(() => {
    if (spy) {
      spy.mockClear();
    }
  });
});
