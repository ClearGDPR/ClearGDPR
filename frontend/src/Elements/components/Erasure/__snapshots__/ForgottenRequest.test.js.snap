// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgottenRequest should render correctly 1`] = `
<Component
  cg={
    CG {
      "Events": Events {
        "_events": Object {},
      },
      "Processors": Processors {
        "_request": RequestFactory {
          "config": [Circular],
        },
      },
      "Subject": Subject {
        "_request": RequestFactory {
          "config": [Circular],
        },
        "eraseData": [Function],
      },
      "_config": Object {
        "apiKey": "test",
        "apiTimeout": 30,
        "apiUrl": "test",
        "apiVersion": undefined,
        "auth": null,
        "debug": false,
      },
    }
  }
  options={
    Object {
      "label": "Erase",
    }
  }
  subject={
    Object {
      "data": Object {},
      "initNewSession": [Function],
      "status": 1,
    }
  }
>
  <ForgottenRequest
    cg={
      CG {
        "Events": Events {
          "_events": Object {},
        },
        "Processors": Processors {
          "_request": RequestFactory {
            "config": [Circular],
          },
        },
        "Subject": Subject {
          "_request": RequestFactory {
            "config": [Circular],
          },
          "eraseData": [Function],
        },
        "_config": Object {
          "apiKey": "test",
          "apiTimeout": 30,
          "apiUrl": "test",
          "apiVersion": undefined,
          "auth": null,
          "debug": false,
        },
      }
    }
    options={
      Object {
        "label": "Erase",
      }
    }
  >
    <button
      className="button is-primary"
      onClick={[Function]}
    >
      Erase
    </button>
    <ModalView
      onClose={[Function]}
      open={false}
    >
      <Modal
        animationDuration={500}
        center={false}
        classNames={
          Object {
            "0": "M",
            "1": "o",
            "2": "d",
            "3": "a",
            "4": "l",
            "5": ".",
            "6": "s",
            "7": "c",
            "8": "s",
            "9": "s",
          }
        }
        classes={
          Object {
            "closeButton": "styles_closeButton__20ID4",
            "closeIcon": "styles_closeIcon__1QwbI",
            "modal": "styles_modal__gNwvD",
            "overlay": "styles_overlay__CLSq-",
            "overlayCenter": "styles_overlayCenter__YHoO7",
            "transitionEnter": "styles_transitionEnter__3j_-a",
            "transitionEnterActive": "styles_transitionEnterActive___eQs7",
            "transitionExit": "styles_transitionExit__1KmEf",
            "transitionExitActive": "styles_transitionExitActive__1nQXw",
          }
        }
        closeIconSize={28}
        closeIconSvgPath={
          <path
            d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
          />
        }
        closeOnEsc={true}
        closeOnOverlayClick={true}
        onClose={[Function]}
        onEntered={null}
        onEscKeyDown={null}
        onExited={null}
        onOverlayClick={null}
        open={false}
        showCloseIcon={true}
        styles={Object {}}
      />
    </ModalView>
  </ForgottenRequest>
</Component>
`;
